portal/server/utils/csv-utils.js

```diff
@@ -41,6 +41,28 @@ const excelColumnToColumnIndex = (column) => {
   return result - 1;
 };
 
+/**
+ * Checks for floating-point rounding errors and rounds to two decimal places if within error tolerance.
+ * @param {number} number - The number to check and potentially round.
+ * @returns {number} - The rounded number to two decimal places if within tolerance, otherwise the original number.
+ * @throws {TypeError} - If the input is not a number.
+ */
+const handleFloatingPointRoundingErrors = (number) => {
+  if (typeof number !== 'number') {
+    throw new TypeError('Input must be a number');
+  }
+
+  const TOLERANCE = 1e-6;
+  const DECIMAL_PLACES = 2;
+
+  const roundedNumber = Number(number.toFixed(DECIMAL_PLACES));
+
+  const isRoundedNumberWithinErrorTolerance = Math.abs(number - roundedNumber) < TOLERANCE;
+
+  return isRoundedNumberWithinErrorTolerance ? roundedNumber : number;
```

Comment:
My final little review comment in passing is why would we want to ever not return the rounded number? I feel this has some odd behaviour as a result?

ie if we had a number that was 1.234 parsed in, this could be the `actual` value in the database, so we'd want to display it as this. Which is great -- this would work fine.

However, if we had performed an operation on this in the database and it had a rounding error (1.2340000001), this would still return 1.2340000001 -- which still has a floating point error.

As we're always wanting 2dp (as its hard coded in) we could just do something like the below maybe?

```
  const handleFloatingPointRoundingErrorsTwo = (number) => {
    if (typeof number !== 'number') {
      throw new TypeError('Input must be a number');
    }
    const DECIMAL_PLACES = 2;

    return Number(number.toFixed(DECIMAL_PLACES));;
  };
```

Reply:
in your PR you mention 'This approach means we can still handle genuine errors such as 12.345 (which isn't 2 DP, but is not a floating-point rounding error either).', but arent these also potentially exhibiting floating point errors due to them being subject to the same calcs that are being done to give the others floating point errors? :)

Reply:
(I think the approach should be that we round to our tolerance and compare that vs the actual number? So `1.1230001` would be rounded to `1.1230000` and the diff would be `0.0000001`), so it would return `1.1230000`.

As we are dealing with such low values however, are we not getting into the territory of the answer having a floating error (so it'd just make sense to just return the rounded and be done with it -- no checks)?

Reply:
Would be keen to catch up with you this morning on this Alex.
I think I kind of understand your point, where maybe we just round to 7 decimal places for example before doing anything else so that if we need to return an error on 1.123 we can say the error is 1.123 not 1.123000001 in the error message. toFixed wouldn't work here but some math.round function might, will have a play in node

Reply:
Spoke to Matt, we're going to do:

Round to x decimal places, truncate 0s, that's now our number in string format that we run our validations on etc.

If it's javascript float precision error then even with extra decimal places we won't round away from them, but if it's a valid entry we will truncate our 0s down to 2 decimal places and we should be happy

Reply:
Omitted the truncating as this is done for us automatically in `parseXlsxToCsvArrays` when we concatenate our cell values together when writing the CSV. Implemented the rounding approach though and updated the tests and pushed this work.

Agree with the end of the previous comment.

This approach still flags genuine errors such as 12.345 during validation (our BA expressed that we want to enforce 2 DP, so we need to keep support for those cases).

Tested locally with problematic report, works as expected.

Reply:
And just for reference, the updated test cases cover a wider range of currency values (hundreds of millions, down to £1.XX), e.g.: `987654321.1200005`, `9999.9999999999`, `1.230001`

Reply:
Spoke to Alex - happy to resolve